packages/dd-trace/src/appsec/user_tracking.js

```diff
@@ -0,0 +1,168 @@
+const { SAMPLING_MECHANISM_APPSEC } = require('../constants')
+const standalone = require('./standalone')
+const waf = require('./waf')
+
+// the RFC doesn't include '_id', but it's common in MongoDB
+const USER_ID_FIELDS = ['id', '_id', 'email', 'username', 'login', 'user']
+
+let collectionMode
+
+function setCollectionMode (mode, overwrite = true) {
+  // don't overwrite if already set, only used in appsec/index.js to not overwrite RC values
+  if (!overwrite && collectionMode) return
+
+  /* eslint-disable no-fallthrough */
+  switch (mode) {
+    case 'safe':
+      log.warn('Using deprecated value "safe" in config.appsec.eventTracking.mode')
+    case 'anon':
+    case 'anonymization':
+      collectionMode = 'anonymization'
+      break
+
+    case 'extended':
+      log.warn('Using deprecated value "extended" in config.appsec.eventTracking.mode')
+    case 'ident':
+    case 'identification':
+      collectionMode = 'identification'
+      break
+
+    default:
+      collectionMode = 'disabled'
+  }
+  /* eslint-enable no-fallthrough */
+}
+
+function obfuscateIfNeeded (str) {
+  if (collectionMode === 'anonymization') {
+    // get first 16 bytes of sha256 hash in lowercase hex
+    return 'anon_' + crypto.createHash('sha256').update(str).digest().toString('hex', 0, 16).toLowerCase()
+  } else {
+    return str
+  }
+}
+
+// TODO: should we find other ways to get the user ID ?
+function getUserId (user) {
+  if (!user) return
+
+  for (const field of USER_ID_FIELDS) {
+    let id = user[field]
+
+    // try to find a field that can be stringified
+    if (id && typeof id.toString === 'function') {
+      id = id.toString()
+
+      if (typeof id !== 'string' || id.startsWith('[object ')) {
+        // probably not a usable ID ?
+        continue
+      }
+
+      return obfuscateIfNeeded(id)
+    }
+  }
+}
+
+function trackLogin (framework, login, user, success, rootSpan) {
+  if (!collectionMode || collectionMode === 'disabled') return
+
+  if (!rootSpan) {
+    log.error('No rootSpan found in AppSec trackLogin')
+    return
+  }
+
+  if (!login || typeof login !== 'string') {
+    log.error('Invalid login provided to AppSec trackLogin')
+
+    telemetry.incrementMissingUserLogin(framework, success ? 'login_success' : 'login_failure')
+    // note:
+    //  if we start supporting using userId if login is missing, we need to only give up if both are missing, and
+    //  implement 'appsec.instrum.user_auth.missing_user_id' telemetry too
+    return
+  }
+
+  login = obfuscateIfNeeded(login)
+  const userId = getUserId(user)
+
+  let newTags
+
+  const persistent = {
+    [addresses.USER_LOGIN]: login
+  }
+
+  const currentTags = rootSpan.context()._tags
+  const isSdkCalled = currentTags[`_dd.appsec.events.users.login.${success ? 'success' : 'failure'}.sdk`] === 'true'
+
+  // used to not overwrite tags set by SDK
+  function shouldSetTag (tag) {
+    return !(isSdkCalled && currentTags[tag])
```

Comment:
```suggestion
    return !isSdkCalled || !currentTags[tag]
```

I think that this is more readable, should set tag when sdk is not called or tag is not set, rather than should be called when not happening these two things: sdk is called and current tag is set.

Reply:
hahahaha honestly this is exactly what i instinctively wrote at first, but when i read it later i got confused and wasn't sure it was the correct logic, the two are logically equivalent but i rewrote it with the parenthesis because it felt more "english" to me:
```
if sdk is called and the tag already exist then do NOT add a new tag

vs

if sdk is not called or the tag doesn't already exist then add a new tag
```
which one feels more natural language ?

Reply:
can i close this ?